### DISCUSSION TITLE
Common fields section when defining schema

### DISCUSSION THREAD

**Discussion by jtaccinelli** [tinacms/tinacms#4034] [Ideas]
Hello,

I've recently started looking at TinaCMS as an alternative for Sanity and am loving it so far. One thing I'm really missing though is that, in Sanity, you can define custom objects as field types, and then reference them in other documents and objects. While I understand that this is a limitation with GraphQL, and that explicit and unique types are required for the purposes of defining the queries, I do think that there is a significant DevX benefit to allowing people to define 'common fields'. 

This is particularly painful when working with types that are super common across all collections - like a `Link` object that might support templates for an external link, and then an internal reference to any one of the documents. The generated types off the back of these kinds of objects are insanely large, and incredibly difficult to work around. I think it'd really reduce the amount of generated Typescript and really help with appropriate typing of components or functions that make use of those types, especially ones where you're expecting the same type to come from any collection, or template.

Would love some thoughts on this, or if it's already in the works, or if it's just impossible!

Thanks.

**Comment by jeffsee55** [comment]
Hi there, I think it's definitely possible and it's not something we've had a lot of requests for. We did have a concept for global templates that may have served this purpose but if proved to be complex because you could recursively reference them. It's something we could have probably put a limit on, but in general I think what you're asking for is reasonable, but maybe not something we'd be able to crank out soon.

One thought I had for this was that maybe we can just allow the developer to provide the GraphQL type name for a given `object` field. And when that object is reused across the schema it's never namespaced, would this be what you had in mind?


```ts
const link = {
  type: "object",
  name: "link",
  label: "Link",
  meta: {
    // This would be normally be namespaced, so it'd be `PageBlocksHeroLink`
    // in one area, and then `PageBlocksNewsletterSignupLink` in another
    graphqlType: "Link",
  },
  templates: [
    {
      name: "referenceLink",
      label: "Reference Link",
      fields: [{ type: "reference", name: "reference", label: "Reference", collections: ['authors'] }],
    },
    // ... various templates here
  ],
};
```
I'm not sure about how much of a lift this would be, but seems doable. It'd be up to the developer to ensure they haven't given a collection the name `link` as that would cause collisions.

**Reply by jtaccinelli** [reply]
Hi - I actually think as an interim measure this would be perfect. 

Having has some more time to look at the inner workings of the thing, I think a really neat solution would actually be to allow people to define _fragments_ globally, with similar syntax to templates, and then let templates and objects extend on those fragments.

Defining them might look like:

```js
defineConfig({
   // Standard config shenanigans,
  schema: {
    collections: [],
    fragments: [
      {
        name: 'link',
        fields: [
          {
            type: 'string',
            name: 'href',
            required: true
          },
          {
            type: 'string',
            name: 'label',
            required: true
          },
        ]
      }
    ]
  }
})
```

And then, creating a link component might look something like: 

```js
const link = {
  type: 'object',
  extends: 'link', // only allows extending from 'fragments'
  name: 'link',
  label: 'Link',
  fields: [
    // Any additional fields,
  ]
}
```

Or, in the case of templates being used:

```js
const link = {
  type: 'object',
  name: 'link',
  label: 'Link',
  templates: [
    {
      name: 'external',
      label: 'External',
      extends: 'link'
    }
  ]
  fields: [
    // Any additional fields,
  ]
}
```

Then, under the hood, those defined object types just adopt the define fragments and add any additional fields on top of that. Then, the typings would be able to use those fragments as interfaces to extend off of.

In GQL, maybe it gets generated as:

```gql
interface Link {
  href: String!,
  label: String!
}

type ExternalLink implements Link {
   // Any additional fields
}

type ExternalLinker {
  link: ExternalLink!
}
```

And then in typescript we get to say:

```ts
interface Link {
 href: string,
 label: string.
}

interface ExternalLink extends Link {
   // Any additional fields
}

type ExternalLinker {
  link: ExternalLink
}
```
Which means I can do relevant typechecks with something like:

```ts
function parseHref<L extends Link>(link: L) {
   // whatever I need to do here
}
```

This feels more like a 'TinaCMS' oriented solution instead of just backfilling for a feature that Sanity has. This is also, obviously, a lot more work I'm guessing than what you're suggesting, but I really think this would allow for some super neat things.

I also think it keeps a clearer mental model, where I think if you settled on just allowing people to un-namespace objects, you might get people doing it in places that wouldn't really benefit them at all, or not using it where they should. I think the idea of 'I'm going to be repeating this a bunch so I'm going to add a singularly defined data structure for the thing' keeps a clear definition between an 'Object you're going to be defining to display in the CMS' and a 'Fragment I want to explicitly define because it's used in a lot of places.'

It helps keeps things really, really DRY and also (I think) avoids the issue of recursive references because you can never outright just use the fragment without having to define a base object that makes use of it, so no one fragment is ever going to be able to reference itself, or any _other_ fragments.

Not being able to reference one fragment in another fragment might be a bit of a frustration though, as I think the potential for this would be unlocked by that, but who knows.

I'm not really a professional in this space, but I'd be really keen to hear your thoughts on this!

**Comment by HenrySkup** [comment]
Hi, have just done a big push with Tina (and Hugo) and some sort of solution in this space would be very helpful as I seemed to be hitting some recursive-reference limits.  I think it would be very helpful to be able to reference data fragments/whatever to help both the UI and the data schemas consistent/memorable -- and more maintainable.
